File: 22-GenerateParentheses/22-GenerateParentheses.java
// Last updated: 16/10/2025, 04:31:17
class Solution {

    public void storeParenthesis(String str, int open, int close, int n, List<String> ans){
        if(open == n && close == n){
            ans.add(str);
            return;
        }

        if(open > n || close > n){
            return;
        }

        if(close > open){
            return;
        }

        storeParenthesis(str + "(", open + 1, close, n, ans);
        storeParenthesis(str + ")", open, close + 1, n, ans);
    }

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        storeParenthesis("", 0, 0, n, ans);

        return ans;
    }
}

File: 22-GenerateParentheses/22-GenerateParentheses.cpp
// Last updated: 16/10/2025, 04:31:24
class Solution {
public:
    
     void helper(string str, int open, int close, vector<string> &ans, int n){
        if(open == n && close == n){
            ans.push_back(str);
            return;
        }
        if(open > n){
            return;
        }
        helper(str + "(", open+1, close, ans, n);
        
        if(open > close){
            helper(str + ")", open, close+1, ans, n);
        }
    }
    
    vector<string> generateParenthesis(int n) 
    {
        vector<string> ans;
        helper("",0,0,ans,n);
        return ans;
    }
};